File: DSA/Pranvkumar_590011587/Day55_Question3.py
def num_subarrays_with_sum(nums, goal):
    from collections import defaultdict
    prefix_counts = defaultdict(int)
    prefix_counts[0] = 1
    curr_sum = 0
    count = 0
    for num in nums:
        curr_sum += num
        if (curr_sum - goal) in prefix_counts:
            count += prefix_counts[curr_sum - goal]
        prefix_counts[curr_sum] += 1
    return count

# Example
print(num_subarrays_with_sum([1,0,1,0,1], 2))  
print(num_subarrays_with_sum([0,0,0,0,0], 0))  

File: DSA/pranavdixit_590014418/Day_55_Question_3.c
#include <stdio.h>

#define MAX 1000

typedef struct {
    int sum;
    int freq;
} Entry;

int getFreq(Entry* map, int size, int sum) {
    for (int i = 0; i < size; i++)
        if (map[i].sum == sum)
            return map[i].freq;
    return 0;
}

void addFreq(Entry* map, int* size, int sum, int value) {
    for (int i = 0; i < *size; i++) {
        if (map[i].sum == sum) {
            map[i].freq += value;
            return;
        }
    }
    map[*size].sum = sum;
    map[*size].freq = value;
    (*size)++;
}

int numSubarraysWithSum(int nums[], int n, int goal) {
    Entry map[MAX];
    int size = 0;
    int sum = 0, count = 0;

    addFreq(map, &size, 0, 1); // Base case

    for (int i = 0; i < n; i++) {
        sum += nums[i];
        count += getFreq(map, size, sum - goal);
        addFreq(map, &size, sum, 1);
    }

    return count;
}

int main() {
    int nums[] = {1, 0, 1, 0, 1};
    int goal = 2;
    int n = sizeof(nums) / sizeof(nums[0]);

    printf("Number of subarrays: %d\n", numSubarraysWithSum(nums, n, goal)); // Output: 4
    return 0;
}
